File: python_client/storage/manifest_store.py
import json
from pathlib import Path


class ManifestStore:

    def __init__(self, path: Path):
        self.path = path

    def _load_all(self) -> dict:
        if not self.path.exists():
            return {}
        return json.loads(self.path.read_text())

    def save(self, manifest: dict):
        """Save a single manifest keyed by its file_name."""
        all_manifests = self._load_all()
        filename = manifest.get("file_name")
        if filename:
            all_manifests[filename] = manifest
        self.path.parent.mkdir(parents=True, exist_ok=True)
        self.path.write_text(json.dumps(all_manifests, indent=2))

    def get(self, filename: str) -> dict | None:
        """Retrieve the manifest for a specific file, or None if not found."""
        return self._load_all().get(filename)

    def list_all(self) -> dict:
        """Return all stored manifests keyed by filename."""
        return self._load_all()

File: java_client/src/main/java/com/cisc468share/storage/ManifestStore.java
package com.cisc468share.storage;

import com.cisc468share.protocol.CanonicalJson;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.nio.file.Files;
import java.nio.file.Path;
import java.util.LinkedHashMap;
import java.util.Map;

/**
 * JSON-backed manifest storage keyed by filename.
 * Persists to data/manifests.json.
 */
public class ManifestStore {

    private final Path path;
    private final ObjectMapper mapper = CanonicalJson.mapper();

    public ManifestStore(Path path) {
        this.path = path;
    }

    private Map<String, Map<String, Object>> loadAll() {
        if (!Files.exists(path)) return new LinkedHashMap<>();
        try {
            return mapper.readValue(path.toFile(),
                    new TypeReference<Map<String, Map<String, Object>>>() {});
        } catch (Exception e) {
            return new LinkedHashMap<>();
        }
    }

    /** Save a manifest keyed by its file_name field. */
    public void save(Map<String, Object> manifest) {
        String filename = (String) manifest.get("file_name");
        if (filename == null) return;
        Map<String, Map<String, Object>> all = loadAll();
        all.put(filename, manifest);
        write(all);
    }

    /** Retrieve manifest for a filename, or null if not found. */
    public Map<String, Object> get(String filename) {
        return loadAll().get(filename);
    }

    /** Return all stored manifests. */
    public Map<String, Map<String, Object>> listAll() {
        return loadAll();
    }

    private void write(Map<String, Map<String, Object>> data) {
        try {
            Files.createDirectories(path.getParent());
            mapper.writerWithDefaultPrettyPrinter().writeValue(path.toFile(), data);
        } catch (Exception e) {
            System.out.println("[ERROR] Failed to write manifest store: " + e.getMessage());
        }
    }
}
